src/v/cluster/archival_metadata_stm.cc

```diff
@@ -96,17 +110,39 @@ archival_metadata_stm::segments_from_manifest(
           return s1.meta.base_offset < s2.meta.base_offset;
       });
 
+    for (auto [key, meta] : manifest) {
+        if (meta.ntp_revision == model::initial_revision_id{}) {
+            meta.ntp_revision = manifest.get_revision_id();
+        }
+        auto name = cloud_storage::generate_local_segment_name(
+          key.base_offset, key.term);
+        segments.push_back(segment{
+          .ntp_revision_deprecated = meta.ntp_revision,
+          .name = std::move(name),
+          .meta = meta});
+    }
+
+    // NOTE: replaced segments shoud go fist in the list
+    // so they will be replaced by the actual segments in
+    // the new manifest.
+    auto sort_begin = segments.begin();
+    std::advance(sort_begin, replaced.size());
+    std::sort(sort_begin, segments.end(), [](const auto& s1, const auto& s2) {
+        return s1.meta.base_offset < s2.meta.base_offset;
+    });
```

Comment:
question: would this section of `segments` (`sort_begin` -> `segments.end()`) be already sorted, as we created this by iterating over an ordered map?

Reply:
makes sense, I think that I just followed the original code